File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        //first we need to build the left and right subarrays
        vector<int> left(height.size());
        vector<int> right(height.size());

        //build left array
        int maxLeft = 0;
        for (int i = 0; i < height.size(); i++) {
            left[i] = max(height[i], maxLeft);
            if (maxLeft < height[i]) maxLeft = height[i];
        }

        //build right array
        int maxRight = 0;
        for (int i = height.size() - 1; i >= 0; i--) {
            right[i] = max(height[i], maxRight);
            if (maxRight < height[i]) maxRight = height[i];
        }

        int totalWater = 0;
        for (int i = 0; i < height.size(); i++) {
            totalWater += min(right[i], left[i]) - height[i];
        }
        return totalWater;
        
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        if len(s) == 1:
            return 1
        
        book = {}
        right = 0
        left = 0
        final_max = 0


        while right < len(s):
            if s[right] not in book:
                book[s[right]] = 1
            else:
                book[s[right]] += 1
            
            right += 1
            window_rn = right - left
            book_max = max(book.values())

            if window_rn - book_max <= k:
                final_max = max(final_max, window_rn)
            else:
                book[s[left]] -= 1
                left += 1
            
        return final_max
            


        